stack-graphs/src/assert.rs

```diff
@@ -72,6 +72,22 @@ impl AssertionSource {
                     .unwrap_or(false)
         })
     }
+
+    pub fn display<'a>(&'a self, graph: &'a StackGraph) -> impl std::fmt::Display + 'a {
+        struct Displayer<'a>(&'a AssertionSource, &'a StackGraph);
+        impl std::fmt::Display for Displayer<'_> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(
+                    f,
+                    "{}:{}:{}",
+                    self.1[self.0.file],
+                    self.0.position.line + 1,
+                    self.0.position.column.grapheme_offset + 1
+                )
+            }
+        }
+        Displayer(self, graph)
+    }
```

Comment:
You might be right, that we could capture the graph in the `AssertionSource` itself. Without having looked at all the current uses, I think this might make it a lot less flexible to use, though. As I recall, these are constructed from tests _before_ the graph even exists (and their construction does not need a graph either). So capturing the graph earlier would restrict users quite a bit.

In one of the later following PRs I've dropped the use of `AssertionSource` in the CLI again, because it requires full LSP positions which require reading the source file to be calculated.

Reply:
I looked in the source, and it is not exactly as I said. The graph must already exist, because we store a `Handle<File>` in `AssertionSource`. But I think my point is still valid, because capturing the graph would e.g. not allow any `&mut` references as long as the `AssertionSource` exists. In other words, it would still force a user to have the graph fully constructed before creating `AssertionSource`s.

Reply:
That's a great point that having a reference to the `graph` in `AssertionSource` would complicate `&mut` references elsewhere. I still think the current solution is 👍 and a nice way to handle this.